### PR TITLE
Commit for AED PDP, new AED knob and AED test fixes (new PR)

### PR DESCRIPTION
**Pull request by fkeceli**
- Implemented the AED progressive dynamic-programming table optimization implemented in StringEditPeriodicityDetector class which speeds up the algorithm without changing its output. (Formerly known as online AED)
- Added a new knob in the StringEditEpochRecordFilter class: min_detectable_period.
- Some AED related tests were not included in the test/Makefile. Adding these tests to the Makefile exposed broken tests, fixed those tests.

### CODE REVIEW

File: src/EditDistPeriodicityDetector.cpp [hunk 1]
```diff
@@ -55,60 +59,89 @@ namespace geopm
...
+        m_DP[((ii % m_history_buffer_size)  * m_history_buffer_size
+            + (jj % m_history_buffer_size)) * m_history_buffer_size
+            + (mm % m_history_buffer_size)] = val;
+    }
+
+    uint32_t EditDistPeriodicityDetector::Dget(int ii, int jj, int mm) {
+        if (ii <= m_record_count - m_history_buffer_size) {
+            // This value is supposed to be INF but not so large that it gets wrapped around when
+            // a small value is added to it.
+            return std::numeric_limits<uint32_t>::max()/2;
```
Comment (dianarg) [inline src/EditDistPeriodicityDetector.cpp]: could still be a named constant if this is used in a few places

**Review by dianarg — COMMENTED** [review]
Consider bringing back the constant for very large value.  Otherwise looks fine to me

File: src/EditDistPeriodicityDetector.cpp (hunk 1 above)
Comment (fkeceli) [reply src/EditDistPeriodicityDetector.cpp]: It is only used in this function so I think this is fine: less global constants to keep track of.

Comment (cmcantalupo) [reply src/EditDistPeriodicityDetector.cpp]: Avoid early returns, and define constant at top of function for repeated use within the function please.

File: src/EditDistPeriodicityDetector.cpp [hunk 2]
```diff
@@ -55,60 +59,89 @@ namespace geopm
...
+            for (int ii = std::max({1, m_record_count - m_history_buffer_size}); ii < mm + 1; ++ii) {
+                int term = 2;
+                if ((m_record_count - (ii - 1) <= num_recs_in_hist) &&
+                    (m_history_buffer.value(num_recs_in_hist - (m_record_count - (ii - 1))) ==
+                     m_history_buffer.value(num_recs_in_hist - 1)))
+                {
+                    term = 0;
                 }
+                Dset(ii, m_record_count - mm, mm,
+                        std::min({Dget(ii - 1, m_record_count - mm    , mm) + 1,
```
Comment (cmcantalupo) [inline src/EditDistPeriodicityDetector.cpp]: use a intermediate variable to store result of the min() computation.

File: src/EditDistPeriodicityDetector.cpp [hunk 3]
```diff
@@ -55,60 +59,89 @@ namespace geopm
     {
         if (record.event == EVENT_REGION_ENTRY) {
             m_history_buffer.insert(record.signal);
+            ++m_record_count;
             calc_period();
         }
     }
 
+    void EditDistPeriodicityDetector::Dset(int ii, int jj, int mm, uint32_t val) {
+        m_DP[((ii % m_history_buffer_size)  * m_history_buffer_size
```
Comment (cmcantalupo) [inline src/EditDistPeriodicityDetector.cpp:68]: ((ii % m_history_buffer_size) * m_history_buffer_size * m_history_buffer_size

File: src/EditDistPeriodicityDetector.cpp [hunk 4]
```diff
@@ -55,60 +59,89 @@ namespace geopm
...
+            // This value is supposed to be INF but not so large that it gets wrapped around when
+            // a small value is added to it.
+            return std::numeric_limits<uint32_t>::max()/2;
+        }
+        if (mm <= m_record_count - m_history_buffer_size) {
+            // This value is supposed to be INF but not so large that it gets wrapped around when
+            // a small value is added to it.
+            return std::numeric_limits<uint32_t>::max()/2;
+        }
+        return m_DP[((ii % m_history_buffer_size)  * m_history_buffer_size
```
Comment (cmcantalupo) [inline src/EditDistPeriodicityDetector.cpp]: ((ii % m_history_buffer_size) * m_history_buffer_size * m_history_buffer_size

File: src/EditDistPeriodicityDetector.cpp [hunk 5]
```diff
@@ -55,60 +59,89 @@ namespace geopm
...
-                                get_history_value(mm + jj - 1)) ?
-                               2 : 0;
-                    DD[ii][jj][mm] = std::min({DD[ii - 1][jj    ][mm] + 1,
-                                               DD[ii    ][jj - 1][mm] + 1,
-                                               DD[ii - 1][jj - 1][mm] + term});
+        for (int mm = std::max({1, m_record_count - m_history_buffer_size}); mm < m_record_count; ++mm) {
+            for (int ii = std::max({1, m_record_count - m_history_buffer_size}); ii < mm + 1; ++ii) {
+                int term = 2;
+                if ((m_record_count - (ii - 1) <= num_recs_in_hist) &&
+                    (m_history_buffer.value(num_recs_in_hist - (m_record_count - (ii - 1))) ==
```
Comment (cmcantalupo) [inline src/EditDistPeriodicityDetector.cpp]: some local paramters to break this line up would make it easier to read.

**Review by cmcantalupo — CHANGES_REQUESTED** [review]
Some nits, but the 3 dimensional array indexing seems incorrect, and I'm a little confused why that isn't breaking any tests.

File: src/EditDistPeriodicityDetector.cpp (hunk 5 above)
Comment (fkeceli) [reply src/EditDistPeriodicityDetector.cpp]: This is done.

File: src/EditDistPeriodicityDetector.cpp (hunk 1 above)
Comment (fkeceli) [reply src/EditDistPeriodicityDetector.cpp]: Fixed this.

File: src/EditDistPeriodicityDetector.cpp (hunk 2 above)
Comment (fkeceli) [reply src/EditDistPeriodicityDetector.cpp]: Done.

File: src/EditDistPeriodicityDetector.cpp (hunk 4 above)
Comment (fkeceli) [reply src/EditDistPeriodicityDetector.cpp]: Wrote these indices more explicitly to easier understand.

**Review by cmcantalupo — APPROVED** [review]

**Pull request merged**